uamqp/aio/_connection_async.py

```diff
@@ -115,10 +115,11 @@ async def _set_state(self, new_state):
             return
         previous_state = self.state
         self.state = new_state
-        _LOGGER.info("Connection '%s' state changed: %r -> %r", self.container_id, previous_state, new_state)
-        async with create_task_group() as tg:
-            for session in self.outgoing_endpoints.values():
-                await tg.spawn(session._on_connection_state_change)
+        _LOGGER.info("Connection '%s' state changed: %r -> %r", self._container_id, previous_state, new_state)
+        futures = []
+        for session in self.outgoing_endpoints.values():
+            futures.append(asyncio.create_task(session._on_connection_state_change()))
+        await asyncio.gather(*futures)
```

Comment:
What happens in the case of an error here? If one session fails to update state, will that raise?

Reply:
What about in the case of an async cancellation?

Reply:
I tried the following code and it would raise error `asyncio.exceptions.CancelledError`.

```python
import asyncio

async def func2():
    await asyncio.sleep(5)
    raise ValueError('Error')

async def func3(task):
    await asyncio.sleep(3)
    task.cancel()

async def main():
    task1 = asyncio.create_task(func2())
    task2 = asyncio.create_task(func3(task1))
    await asyncio.gather(task1, task2)

asyncio.run(main())
```

Reply:
Interesting that it raised a `CancelledError` - according to the docs:
`the first raised exception is immediately propagated to the task that awaits on gather`
https://docs.python.org/3/library/asyncio-task.html#asyncio.gather
Though either way it looks like errors should be propagated - do we need and error handling for failures in `on_connection_state_changed`?

Lastly - this could be simplified to:
`await asyncio.gather(*[s._on_connection_state_change() for s in self.outgoing_endpoints.values()])`

(List comprehensions will always be more performant than traditional loop syntax ;) )

Reply:
> Though either way it looks like errors should be propagated - do we need and error handling for failures in on_connection_state_changed?

I honestly don't have a clear answer for it.. just taking a look at the code `on_connection_state_changed`, the chain of call is like
1. connection notifies its all subsidiary sessions
2. session notify its all subsidiary links that session state is changed
3. link gets notified and could perform
3.1 attach
3.2 detach and clean up deliveries

I think we should apply single responsibility principle here -- connection module only deals with connection error, error that doesn't belong to this module, leave that error to upstream/downstream to handle..

with that being said, right now, I don't feel we need to handle error here, this is for notifying the upstream that connection state is changed, and upstream modules should take care of themselves.

Reply:
(and thanks for the hint of List comprehensions!)

Reply:
There are quite a few places (where ever we're currently calling `gather`) that should be replaced with list comprehensions for performance - but we can fix that up later.